adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/api/rest/v1/DatasourceEndpoint.java

```diff
@@ -20,52 +22,48 @@
 public class DatasourceEndpoint {
   private final DatasourceManager datasourceManager;
 
+  private <T> T handleErrors(CheckedSupplier<T> function) throws ResponseStatusException {
+    try {
+      return function.get();
+    } catch (Exception e) {
+      HttpStatus errorCode = Mappings.ERROR_MAPPING.getOrDefault(e.getClass(), HttpStatus.INTERNAL_SERVER_ERROR);
+      throw new ResponseStatusException(errorCode, e.getMessage());
+    }
+  }
+
   @GetMapping
   public Iterable<Datasource> getDatasources() {
     return datasourceManager.getAllDatasources();
   }
 
   @GetMapping("/{id}")
   public Datasource getDatasource(@PathVariable Long id) {
-    return datasourceManager.getDatasource(id)
-      .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Could not find datasource with id " + id));
+    return this.handleErrors(() -> datasourceManager.getDatasource(id));
   }
 
   @PostMapping
   public ResponseEntity<Datasource> addDatasource(@Valid @RequestBody Datasource config) {
-    if (config.getId() != null) {
-      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Id is defined by the server. Id field must not be set");
-    }
-    Datasource savedConfig = datasourceManager.createDatasource(config);
-
-    URI location = ServletUriComponentsBuilder.fromCurrentRequest()
+    return this.handleErrors(() -> {
+      Datasource savedConfig = datasourceManager.createDatasource(config);
+      URI location = ServletUriComponentsBuilder.fromCurrentRequest()
       .path("/{id}")
       .buildAndExpand(savedConfig.getId())
       .toUri();
-
-    return ResponseEntity.created(location).body(savedConfig);
+      
+      return ResponseEntity.created(location).body(savedConfig);
+    });
   }
 
   @PutMapping("/{id}")
   @ResponseStatus(HttpStatus.NO_CONTENT)
-  public void updateDatasource(
-    @PathVariable Long id,
-    @Valid @RequestBody Datasource updateConfig) {
-    try {
-      datasourceManager.updateDatasource(id, updateConfig);
-    } catch (IllegalArgumentException e) {
-      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Datasource needs to exist before updating", e);
-    }
+  public void updateDatasource(@PathVariable Long id, @Valid @RequestBody Datasource updateConfig) {
+    this.handleErrors(() -> { datasourceManager.updateDatasource(id, updateConfig); return null;});
```

Comment:
the `return null;` is necessary as the `handleErrors` function expects a `CheckedSupplier` which returns something.

It would be an option to overload that function expecting a (to be defined) `CheckedRunnable` that returns void, but that would mean more code duplication. It could have been so pretty :cry: ...